src/python-bindings/classad2/classad.cpp

```diff
@@ -1 +1,713 @@
+ExprTree * convert_python_to_classad_exprtree(PyObject * py_v);
 
+static PyObject *
+_classad_init( PyObject *, PyObject * args ) {
+    // _classad_init( self, self._handle )
+
+    PyObject * self = NULL;
+    PyObject_Handle * handle = NULL;
+    if(! PyArg_ParseTuple( args, "OO", & self, (PyObject **)& handle )) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    handle->t = new classad::ClassAd();
+    handle->f = [](void * & v){ dprintf( D_PERF_TRACE, "[ClassAd]\n" ); delete (classad::ClassAd *)v; v = NULL; };
+    Py_RETURN_NONE;
+}
+
+
+static PyObject *
+_classad_init_from_string( PyObject *, PyObject * args ) {
+    // _classad_init( self, self._handle, str )
+
+    PyObject * self = NULL;
+    PyObject_Handle * handle = NULL;
+    const char * from_string = NULL;
+    if(! PyArg_ParseTuple( args, "OOz", & self, (PyObject **)& handle, & from_string)) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+    handle->f = [](void * & v){ dprintf( D_PERF_TRACE, "[unconstructed ClassAd]\n" ); if( v != NULL ) { dprintf(D_ALWAYS, "Error!  Unconstructed ClassAd has non-NULL handle %p\n", v); } };
+
+    classad::ClassAdParser parser;
+    classad::ClassAd * result = parser.ParseClassAd(from_string);
+    if( result == NULL ) {
+        // This was ClassAdParseError in version 1.
+        PyErr_SetString( PyExc_SyntaxError, "Unable to parse string into a ClassAd." );
+        return NULL;
+    }
+
+    // I _really_ hope the parser doesn't think it owns the result,
+    // but the CopyFrom() here was in version 1, and removing changes
+    // the output order from the pretty-printer.  That's probably two
+    // bugs (CopyFrom() inserting in the wrong order and the pretty-
+    // printer not sorting), but I really don't feel like fixing either.
+    handle->t = new ClassAd();
+    ((ClassAd *)handle->t)->CopyFrom(* result);
```

Comment:
The copy is not necessary. The ClassAd stores the attributes in a hash table, so insert order doesn't matter. From a brief look at the code, the pretty-printer doesn't sort. But if you call it with a white-list of attributes, they will end up sorted as a side-effect of how they're filtered. I don't mind the code staying for this PR, but we should revisit it later.

Reply:
I agree that the copy is not necessary, but (somehow) the process of making the copy changed the order of the attributes generated by the pretty-printer, and that broke something.  Adding to HTCONDOR-2163.